File: NeetCode/Swift/BinaryTrees/Easy/94.BinaryTreeInorder.py
# Given the root of a binary tree, return the inorder traversal of its nodes' values.

class TreeNode(object):
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

def inorderTraversal(root):
    result, stack = [], []
    current = root

    while current or stack:
        while current:
            stack.append(current)
            current = current.left

        current = stack.pop()
        result.append(current.val)
        current = current.right

    return result


def inorderTraversalR(root):
    result = []

    def helper(root):
        if not root:
            return

        helper(root.left)
        result.append(root.val)
        helper(root.right)

    helper(root)
    
    return result

File: swift/Neetcode/Trees/Easy/94.BinaryTreeInorder.swift
// Given the root of a binary tree, return the inorder traversal of its nodes' values.

// Time: O(n), Space: O(h)
func inorderTraversal(_ root: TreeNode?) -> [Int] {
    var result = [Int]()
    var stack = [TreeNode?]()
    var current = root

    while !stack.isEmpty || current != nil {
        while current != nil {
            stack.append(current)
            current = current?.left
        }

        current = stack.removeLast()
        result.append(current?.val ?? 0)
        current = current?.right
    }
    
    return result
}

func inorderTraversal(_ root: TreeNode?) -> [Int] {
    var result = [Int]()

    func dfs(_ root: TreeNode?) {
        guard let root = root else { return }

        dfs(root.left)
        result.append(root.val)
        dfs(root.right)
    }

    dfs(root)
    return result
}